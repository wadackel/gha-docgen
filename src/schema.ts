import { z } from 'zod';

/**
 * Action Metadata
 * @see https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions
 */
export type ActionInputEntry = z.infer<typeof actionInputEntrySchema>;
export const actionInputEntrySchema = z.object({
  description: z.string(),
  required: z.boolean().optional(),
  default: z.string().optional(),
  deprecationMessage: z.string().optional(),
});

export type ActionInputs = z.infer<typeof actionInputsSchema>;
export const actionInputsSchema = z.record(z.string(), actionInputEntrySchema);

export type ActionOutputEntry = z.infer<typeof actionOutputEntrySchema>;
export const actionOutputEntrySchema = z.object({
  description: z.string(),
});

export type ActionOutputs = z.infer<typeof actionOutputsSchema>;
export const actionOutputsSchema = z.record(z.string(), actionOutputEntrySchema);

export type ActionBranding = z.infer<typeof actionBrandingSchema>;
export const actionBrandingSchema = z
  .object({
    color: z.string(),
    icon: z.string(),
  })
  .partial();

export type Action = z.infer<typeof actionSchema>;
export const actionSchema = z.object({
  name: z.string(),
  author: z.string().optional(),
  description: z.string(),
  inputs: actionInputsSchema.optional(),
  outputs: actionOutputsSchema.optional(),
  branding: actionBrandingSchema.optional(),
});

/**
 * Docgen
 */
export type DocgenStyle = z.infer<typeof docgenStyleSchema>;
export const docgenStyleSchema = z.union([
  z.literal('section:h1'),
  z.literal('section:h2'),
  z.literal('section:h3'),
  z.literal('section:h4'),
  z.literal('section:h5'),
  z.literal('section:h6'),
  z.literal('table'),
]);
