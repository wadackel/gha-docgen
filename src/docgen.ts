import type { Action, DocgenStyle } from './schema';
import { nl2br, nl2md, nl2space } from './utils';

const regex = {
  description: /(<!--\s*gha-description-start\s*-->)([\s\S]*?)(<!--\s*gha-description-end\s*-->)/g,
  inputs: /(<!--\s*gha-inputs-start\s*-->)([\s\S]*?)(<!--\s*gha-inputs-end\s*-->)/g,
  outputs: /(<!--\s*gha-outputs-start\s*-->)([\s\S]*?)(<!--\s*gha-outputs-end\s*-->)/g,
};

const escapeVerticalBar = (input: string) => {
  return input.replace(/\|/g, '\\|');
};

const getSectionHeading = (style: string) => {
  const [, opts] = style.split(':');
  const level = Number(opts!.substring(1));
  return '#'.repeat(level);
};

const inject = (input: string, matcher: RegExp, body: string) => {
  let output = input;
  let match: RegExpExecArray | null;

  while ((match = matcher.exec(output)) !== null) {
    const index = match.index;
    const start = match[1]!;
    const original = match[2]!;
    const end = match[3]!;
    const before = output.substring(0, index);
    const after = output.substring(index + start.length + original.length + end.length);
    output = [before, start, `\n${body}\n`, end, after].join('');
  }

  return output;
};

export type DocgenOptions = {
  style: DocgenStyle;
};

const generateInputsBody = (action: Action, { style }: DocgenOptions) => {
  if (action.inputs === undefined) {
    return '';
  }

  switch (style) {
    case 'section:h1':
    case 'section:h2':
    case 'section:h3':
    case 'section:h4':
    case 'section:h5':
    case 'section:h6': {
      const heading = getSectionHeading(style);
      const list: string[] = [];
      for (const [id, input] of Object.entries(action.inputs)) {
        const required = input.required ? '`true`' : '`false`';
        const defaults = input.default !== undefined ? `\`${input.default}\`` : 'n/a';
        const deprecated = input.deprecationMessage !== undefined ? nl2space(input.deprecationMessage.trim()) : null;
        const parts = [`${heading} \`${id}\``, ``, `**Required:** ${required}  `];
        if (deprecated !== null) {
          parts.push(`**Default:** ${defaults}  `, `**Deprecated:** :warning: ${deprecated}`);
        } else {
          parts.push(`**Default:** ${defaults}`);
        }
        parts.push('', nl2md(input.description.trim()));
        list.push(parts.join('\n'));
      }
      return list.join('\n\n');
    }

    case 'table': {
      // prettier-ignore
      const rows = [
        '| ID | Required | Default | Description |',
        '| :-- | :-- | :-- | :-- |',
      ];
      for (const [id, input] of Object.entries(action.inputs)) {
        const required = input.required ? ':white_check_mark:' : '';
        const defaults = input.default !== undefined ? `\`${input.default}\`` : 'n/a';
        let body = nl2br(input.description.trim());
        if (input.deprecationMessage !== undefined) {
          body += `<br />:warning: **Deprecated:** ${nl2br(input.deprecationMessage.trim())}`;
        }
        const row = [`\`${id}\``, required, defaults, body].map(escapeVerticalBar);
        rows.push(`| ${row.join(' | ')} |`);
      }
      return rows.join('\n');
    }
  }
};

const generateOutputsBody = (action: Action, { style }: DocgenOptions) => {
  if (action.outputs === undefined) {
    return '';
  }

  switch (style) {
    case 'section:h1':
    case 'section:h2':
    case 'section:h3':
    case 'section:h4':
    case 'section:h5':
    case 'section:h6': {
      const heading = getSectionHeading(style);
      const list: string[] = [];
      for (const [id, output] of Object.entries(action.outputs)) {
        const parts = [`${heading} \`${id}\``, ``, nl2md(output.description.trim())];
        list.push(parts.join('\n'));
      }
      return list.join('\n\n');
    }

    case 'table': {
      // prettier-ignore
      const rows = [
        '| ID | Description |',
        '| :-- | :-- |',
      ];
      for (const [id, output] of Object.entries(action.outputs)) {
        const body = nl2br(output.description.trim());
        const row = [`\`${id}\``, body].map(escapeVerticalBar);
        rows.push(`| ${row.join(' | ')} |`);
      }
      return rows.join('\n');
    }
  }
};

export const docgen = (file: string, action: Action, options: DocgenOptions): string => {
  const description = nl2md(action.description.trim());
  const inputs = generateInputsBody(action, options);
  const outputs = generateOutputsBody(action, options);

  let output = file;
  output = inject(output, regex.description, description);
  output = inject(output, regex.inputs, inputs);
  output = inject(output, regex.outputs, outputs);

  return output;
};
