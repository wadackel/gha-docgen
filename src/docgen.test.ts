import { test, expect, describe } from 'vitest';
import { docgen } from './docgen';
import { createAction } from './schema.factory';
import type { DocgenStyle } from './schema';

describe('basic', () => {
  const action = createAction({
    description: 'line1\nline2',
    inputs: {
      id1: {
        description: 'desc1',
      },
      id2: {
        description: 'desc2',
      },
    },
    outputs: {
      id1: {
        description: 'desc1',
      },
      id2: {
        description: 'desc2',
      },
    },
  });

  const options = {
    style: 'section:h3' as const,
  };

  test('non marker', () => {
    const file = `
# Title

## Inputs

## Outputs
    `.trim();

    expect(docgen(file, action, options)).toMatchSnapshot();
  });

  test('empty', () => {
    const file = `
## Inputs

<!-- gha-inputs-start -->
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
<!-- gha-outputs-end -->
    `.trim();

    expect(docgen(file, createAction({}), options)).toMatchSnapshot();
  });

  test('description', () => {
    const file = `
# Title

Êó•Êú¨Ë™û„Å®ÁµµÊñáÂ≠óüìù

<!-- gha-description-start -->
<!-- gha-description-end -->

## Features
    `.trim();

    expect(docgen(file, action, options)).toMatchSnapshot();
  });

  test('multi marker', () => {
    const file = `
# Title

<!--gha-description-start-->
<!--gha-description-end-->

<!--    gha-description-start        -->
<!--
  gha-description-end
-->

## Inputs

<!--gha-inputs-start-->
<!--gha-inputs-end-->

<!-- gha-inputs-start   -->
<!--   gha-inputs-end -->

## Outputs

<!--gha-outputs-start  -->
<!-- gha-outputs-end  -->

<!--gha-outputs-start -->
<!--gha-outputs-end -->
    `.trim();

    expect(docgen(file, action, options)).toMatchSnapshot();
  });
});

describe('style', () => {
  test.each<DocgenStyle>([
    // prettier-ignore
    'section:h1',
    'section:h2',
    'section:h3',
    'section:h4',
    'section:h5',
    'section:h6',
    'table',
  ])('%s', (style) => {
    const action = createAction({
      description: 'description `code`.\nline2'.trim(),
      inputs: {
        id1: {
          description: 'desc1|desc1',
        },
        id2: {
          description: 'desc2\nrequired',
          required: true,
        },
        id3: {
          description: 'desc3',
          deprecationMessage: 'line1\nline2',
        },
        id4: {
          description: 'desc3',
          default: 'https://gha-docgen.example',
        },
      },
      outputs: {
        id1: {
          description: 'desc1|desc1',
        },
        id2: {
          description: 'desc2\nline2',
        },
      },
    });

    const file = `
## Inputs

<!-- gha-inputs-start -->
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
<!-- gha-outputs-end -->
    `.trim();

    expect(docgen(file, action, { style })).toMatchSnapshot();
  });
});
