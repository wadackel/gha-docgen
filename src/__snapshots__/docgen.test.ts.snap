// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > description 1`] = `
"# Title

日本語と絵文字📝

<!-- gha-description-start -->
line1  
line2
<!-- gha-description-end -->

## Features"
`;

exports[`basic > empty 1`] = `
"## Inputs

<!-- gha-inputs-start -->

<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->

<!-- gha-outputs-end -->"
`;

exports[`basic > multi marker 1`] = `
"# Title

<!--gha-description-start-->
line1  
line2
<!--gha-description-end-->

<!--    gha-description-start        -->
line1  
line2
<!--
  gha-description-end
-->

## Inputs

<!--gha-inputs-start-->
### \`id1\`

**Required:** \`false\`  
**Default:** n/a

desc1

### \`id2\`

**Required:** \`false\`  
**Default:** n/a

desc2
<!--gha-inputs-end-->

<!-- gha-inputs-start   -->
### \`id1\`

**Required:** \`false\`  
**Default:** n/a

desc1

### \`id2\`

**Required:** \`false\`  
**Default:** n/a

desc2
<!--   gha-inputs-end -->

## Outputs

<!--gha-outputs-start  -->
### \`id1\`

desc1

### \`id2\`

desc2
<!-- gha-outputs-end  -->

<!--gha-outputs-start -->
### \`id1\`

desc1

### \`id2\`

desc2
<!--gha-outputs-end -->"
`;

exports[`basic > non marker 1`] = `
"# Title

## Inputs

## Outputs"
`;

exports[`style > section:h1 1`] = `
"## Inputs

<!-- gha-inputs-start -->
# \`id1\`

**Required:** \`false\`  
**Default:** n/a

desc1|desc1

# \`id2\`

**Required:** \`true\`  
**Default:** n/a

desc2  
required

# \`id3\`

**Required:** \`false\`  
**Default:** n/a  
**Deprecated:** :warning: line1 line2

desc3

# \`id4\`

**Required:** \`false\`  
**Default:** \`https://gha-docgen.example\`

desc3
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
# \`id1\`

desc1|desc1

# \`id2\`

desc2  
line2
<!-- gha-outputs-end -->"
`;

exports[`style > section:h2 1`] = `
"## Inputs

<!-- gha-inputs-start -->
## \`id1\`

**Required:** \`false\`  
**Default:** n/a

desc1|desc1

## \`id2\`

**Required:** \`true\`  
**Default:** n/a

desc2  
required

## \`id3\`

**Required:** \`false\`  
**Default:** n/a  
**Deprecated:** :warning: line1 line2

desc3

## \`id4\`

**Required:** \`false\`  
**Default:** \`https://gha-docgen.example\`

desc3
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
## \`id1\`

desc1|desc1

## \`id2\`

desc2  
line2
<!-- gha-outputs-end -->"
`;

exports[`style > section:h3 1`] = `
"## Inputs

<!-- gha-inputs-start -->
### \`id1\`

**Required:** \`false\`  
**Default:** n/a

desc1|desc1

### \`id2\`

**Required:** \`true\`  
**Default:** n/a

desc2  
required

### \`id3\`

**Required:** \`false\`  
**Default:** n/a  
**Deprecated:** :warning: line1 line2

desc3

### \`id4\`

**Required:** \`false\`  
**Default:** \`https://gha-docgen.example\`

desc3
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
### \`id1\`

desc1|desc1

### \`id2\`

desc2  
line2
<!-- gha-outputs-end -->"
`;

exports[`style > section:h4 1`] = `
"## Inputs

<!-- gha-inputs-start -->
#### \`id1\`

**Required:** \`false\`  
**Default:** n/a

desc1|desc1

#### \`id2\`

**Required:** \`true\`  
**Default:** n/a

desc2  
required

#### \`id3\`

**Required:** \`false\`  
**Default:** n/a  
**Deprecated:** :warning: line1 line2

desc3

#### \`id4\`

**Required:** \`false\`  
**Default:** \`https://gha-docgen.example\`

desc3
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
#### \`id1\`

desc1|desc1

#### \`id2\`

desc2  
line2
<!-- gha-outputs-end -->"
`;

exports[`style > section:h5 1`] = `
"## Inputs

<!-- gha-inputs-start -->
##### \`id1\`

**Required:** \`false\`  
**Default:** n/a

desc1|desc1

##### \`id2\`

**Required:** \`true\`  
**Default:** n/a

desc2  
required

##### \`id3\`

**Required:** \`false\`  
**Default:** n/a  
**Deprecated:** :warning: line1 line2

desc3

##### \`id4\`

**Required:** \`false\`  
**Default:** \`https://gha-docgen.example\`

desc3
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
##### \`id1\`

desc1|desc1

##### \`id2\`

desc2  
line2
<!-- gha-outputs-end -->"
`;

exports[`style > section:h6 1`] = `
"## Inputs

<!-- gha-inputs-start -->
###### \`id1\`

**Required:** \`false\`  
**Default:** n/a

desc1|desc1

###### \`id2\`

**Required:** \`true\`  
**Default:** n/a

desc2  
required

###### \`id3\`

**Required:** \`false\`  
**Default:** n/a  
**Deprecated:** :warning: line1 line2

desc3

###### \`id4\`

**Required:** \`false\`  
**Default:** \`https://gha-docgen.example\`

desc3
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
###### \`id1\`

desc1|desc1

###### \`id2\`

desc2  
line2
<!-- gha-outputs-end -->"
`;

exports[`style > table 1`] = `
"## Inputs

<!-- gha-inputs-start -->
| ID | Required | Default | Description |
| :-- | :-- | :-- | :-- |
| \`id1\` |  | n/a | desc1\\|desc1 |
| \`id2\` | :white_check_mark: | n/a | desc2<br />required |
| \`id3\` |  | n/a | desc3<br />:warning: **Deprecated:** line1<br />line2 |
| \`id4\` |  | \`https://gha-docgen.example\` | desc3 |
<!-- gha-inputs-end -->

## Outputs

<!-- gha-outputs-start -->
| ID | Description |
| :-- | :-- |
| \`id1\` | desc1\\|desc1 |
| \`id2\` | desc2<br />line2 |
<!-- gha-outputs-end -->"
`;
